lib/HTFeed/StorageZipAudit/ObjectStore.pm

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/perl
+package HTFeed::StorageZipAudit::ObjectStore;
+
+use strict;
+use warnings;
+use Carp;
+use Log::Log4perl qw(get_logger);
+use File::Temp;
+
+use HTFeed::Config qw(get_config);
+use HTFeed::DBTools qw(get_dbh);
+use HTFeed::Volume;
+
+use base qw(HTFeed::StorageZipAudit);
+
+sub storage_object_path {
+  my $self = shift;
+  my $obj  = shift;
+
+  unless ($obj->{tmpdir}) {
+    $obj->{tmpdir} = File::Temp->newdir();
+  }
+  return $obj->{tmpdir};
+}
+
+sub all_objects {
+  my $self = shift;
+
+  $self->_request_object($self->random_object());
+  my $sql = 'SELECT namespace,id,path,version FROM feed_backups' .
+            ' WHERE storage_name=?' .
+            ' AND restore_request IS NOT NULL';
+  $self->{pending_objects} = [];
+  eval {
+    my $ref = get_dbh->selectall_arrayref($sql, undef, $self->{storage_name});
+    foreach my $row (@$ref) {
+      my $obj = $self->storage_object(@$row);
+      if (1 == $self->_restore_object($obj)) {
+        push @{$self->{pending_objects}}, $obj;
+      }
+    }
+  };
+  if ($@) {
+    die "Database query failed: $@";
+  }
+  return $self->{pending_objects};
+}
+
+# Request a single object from Glacier. To keep storage costs down,
+# use the restore_request_issued flag to keep the requests under control
+# in case something else goes haywire.
+sub _request_object {
```

Comment:
I wonder whether this ought to be in `HTFeed::Storage::S3` (or perhaps `HTFeed::Storage::S3::Glacier`)?

Reply:
I'd separate out the SQL `feed_backups` updates, however. So restore request made by some other process doesn't end up as unwilling audit participant.

Reply:
:+1: